### PR TITLE
relax dependency semver range check for inter-repo dependencies

### PR DESCRIPTION
**Pull request by edwardfoyle**
## Problem

With [this change](https://github.com/aws-amplify/amplify-backend/pull/1487), inter-repo dependencies are no longer all on the exact same semver range. This is okay, but caused our dependency_validator check to fail because it detected inconsistent version ranges. eg https://github.com/aws-amplify/amplify-backend/actions/runs/9083855361/job/24963522397?pr=1508#step:5:40

**Issue number, if available:**

## Changes

Relax the dependency_validator to only check that inter-repo dependencies are on a consistent major version rather than exact semver range. External dependencies are still validated to all be on the exact same semver range.

**Corresponding docs PR, if applicable:**

## Validation

Added a unit test and manually verified.

## Checklist

- [x] If this PR includes a functional change to the runtime behavior of the code, I have added or updated automated test coverage for this change.
- [x] If this PR requires a change to the [Project Architecture README](../PROJECT_ARCHITECTURE.md), I have included that update in this PR.
- [x] If this PR requires a docs update, I have linked to that docs PR above.
- [x] If this PR modifies E2E tests, makes changes to resource provisioning, or makes SDK calls, I have run the PR checks with the `run-e2e` label set.

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

**Review by sobolk — APPROVED** [review]

File: scripts/components/dependencies_validator.test.ts [hunk 1]
```diff
@@ -130,7 +130,9 @@ void describe('Dependency validator', () => {
       },
       (err: Error) => {
         assert.ok(
-          err.message.includes('is declared using inconsistent versions')
+          err.message.includes(
+            'dependency declarations must all the on the same semver range'
```
Comment (rtpascual) [inline scripts/components/dependencies_validator.test.ts:134]:
```suggestion
            'dependency declarations must all be on the same semver range'
```
Small nit

File: scripts/components/dependencies_validator.ts [hunk 2]
```diff
@@ -238,4 +239,70 @@ export class DependenciesValidator {
...
+          `${packageName} dependency declarations must all be on the same major version`
+        );
+      };
+    }
+    // default behavior for all other packages is that they must be on the same version
+    return (declarations) => {
+      const versionSet = new Set(declarations.map(({ version }) => version));
+      return (
+        versionSet.size === 1 ||
+        `${packageName} dependency declarations must all the on the same semver range`
```
Comment (rtpascual) [inline scripts/components/dependencies_validator.ts:286]:
```suggestion
        `${packageName} dependency declarations must all be on the same semver range`
```
Same here

File: scripts/components/dependencies_validator.test.ts (hunk 1 above)
Comment (Amplifiyer) [inline scripts/components/dependencies_validator.test.ts:134]: nit:
```suggestion
            'dependency declarations must all be on the same semver range'
```

**Review by rtpascual — APPROVED** [review]
LGTM, just a couple small nits

File: scripts/components/dependencies_validator.test.ts [hunk 3]
```diff
@@ -142,6 +144,33 @@ void describe('Dependency validator', () => {
     );
   });
 
+  void it('can detect inconsistent major versions of repo packages', async () => {
```
Comment (Amplifiyer) [inline scripts/components/dependencies_validator.test.ts:147]: Should we also add a test that inconsistent non-major versions of repo packages are ok?

**Review by 0618 — APPROVED** [review]

**Pull request merged**